### PR TITLE
perf: nocopy read for ReadString and ReadBinary API if possible

### PR DESCRIPTION
**Pull request by joway**
#### What type of PR is this?

perf

#### Check the PR title.

- [x] This PR title match the format: \<type\>(optional scope): \<description\>
- [x] The description of this PR title is user-oriented and clear enough for others to understand.
- [x] Attach the PR updating the user documentation if the current PR requires user awareness at the usage level. [User docs repo](https://github.com/cloudwego/cloudwego.github.io)

#### (Optional) More detailed description for this PR(en: English/zh: Chinese).

en:

before:
```
BenchmarkNoCopyRead-12    	  199310	      5662 ns/op	   32809 B/op	      11 allocs/op
```
after:
```
BenchmarkNoCopyRead-12    	17504086	        80.34 ns/op	      56 B/op	       2 allocs/op
```

zh(optional):

### CODE REVIEW

File: nocopy_linkbuffer.go [hunk 1]
```diff
@@ -220,15 +213,27 @@ func (b *LinkBuffer) ReadBinary(n int) (p []byte, err error) {
...
 
 // readBinary cannot use mcache, because the memory allocated by readBinary will not be recycled.
-func (b *LinkBuffer) readBinary(n int) (p []byte) {
+func (b *UnsafeLinkBuffer) readBinary(n int) (p []byte) {
 	b.recalLen(-n) // re-cal length
 
 	// single node
-	p = make([]byte, n)
 	if b.isSingleNode(n) {
+		// if readBinary use no-copy mode, it will cause more memory used but get higher memory access efficiently
```
Comment (joway) [inline nocopy_linkbuffer.go:221]: 关键代码在这里；其余部分是为了避免每次修改linkbuffer都要改两份做的重构

File: nocopy_linkbuffer.go [hunk 2]
```diff
@@ -220,15 +213,27 @@ func (b *LinkBuffer) ReadBinary(n int) (p []byte, err error) {
...
+		// for example, if user's codec need to decode 10 strings and each have 100 bytes, here could help the codec
+		// no need to malloc 10 times and the string slice could have the compact memory allocation.
+		if !b.read.getMode(reuseMask) {
```
Comment (ppzqh) [inline nocopy_linkbuffer.go:224]: 加一个util方法直接判断是否 reuse？

File: nocopy.go [hunk 3]
```diff
@@ -244,10 +248,50 @@ func NewIOReadWriter(rw ReadWriter) io.ReadWriter {
...
+	block4k  = 4 * 1024
+	block8k  = 8 * 1024
+	block32k = 32 * 1024
+
+	pagesize  = block8k
+	mallocMax = block8k * block1k // mallocMax is 8MB
+
+	defaultLinkBufferMode = 1 << 0
+	minReuseBytes         = 64 // only reuse bytes if n >= minReuseBytes
+	// reuse mode, indicate weather reuse buffer node data, default true
```
[the remark this answers is not in the document]
Comment (ppzqh) [reply nocopy.go:262]: 怎么感觉直接用 nocopyReadMask 或者叫 reused 会更好理解。整体的idea就是nocopyRead，然后 release 的时候不回收对吧

[the remark this answers is not in the document]
Comment (ppzqh) [reply nocopy.go:262]: reusable

File: nocopy_linkbuffer.go (hunk 2 above)
Comment (joway) [reply nocopy_linkbuffer.go:224]: 没必要吧？没多少地方用到这个东西

Comment (ppzqh) [reply nocopy_linkbuffer.go]: ok

**Review by ppzqh — DISMISSED** [review]

**Review by jayantxie — DISMISSED** [review]

**Review by jayantxie — APPROVED** [review]

**Review by ppzqh — APPROVED** [review]

File: nocopy_linkbuffer.go [hunk 4]
```diff
@@ -285,9 +293,9 @@ func (b *LinkBuffer) Slice(n int) (r Reader, err error) {
 	b.recalLen(-n) // re-cal length
 
 	// just use for range
-	p := &LinkBuffer{
-		length: int64(n),
-	}
+	p := new(LinkBuffer)
```
Comment (kingcanfish) [inline nocopy_linkbuffer.go:296]: May I ask, what are the differences between these two approaches~ thanks
i think they are same?
```
p := &LinkBuffer{
		length: int64(n),
	}
```
```
p := new(LinkBuffer)
p.length = int64(n)
```

Comment (joway) [reply nocopy_linkbuffer.go:296]: @kingcanfish no different just to make code more clean

**Pull request merged**